Updates for CMIP7

dabail10 opened the pull request:
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>


## PR checklist
- [X] Short (1 sentence) summary of your PR: 
Starting this in draft mode. This is a big PR with all of the SIMIP updates for CMIP7.
- [X] Developer(s): 
dabail10 (D. Bailey)
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
 io_suite results here
https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#6b7cd825797762e22fe0a05342917d574a980966
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.

This PR should address issue #1048 and #1038.

- I have updated all of the variable names, long names, and units to correspond to the CMIP7 data request.
- I have added new variables requested in the CMIP7 data request.
- I have added documentation about the CMIP6 to CMIP7 update.
- I have simplified the accumulation of some fields where possible and added prognostic sea ice density.
- I have added accumulation of variables relative to aice_init or aice.
- Bug fix for flwout (sifllwutop) where aice_init = 0, but aice > 0.
- Bug fix for shortwave abosrbed and albedo computation (more coming later)
- Bug fix: Some variables that were scaled by aice, should be multiplied by aice (not aice_init) to get the _ai quantities. This includes fswabs, fsens, flat, ...
- Remove f_CMIP flag and add set_nml.cmip option instead.
- Added comment field for SIMIP variables that uses part of the description field in the CMIP data request table.
- Add long_name field for time_bounds, lon?_bounds and lat?_bounds. Issue #1057 
- Partly address aice versus aice_init #1033
- Partial fix for albedo variables #1051
- Also addresses issue #904 
- Updated test results

Closes #904
Closes #1033 
Closes #1038 
Closes #1048 
Closes #1057

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1808,15 +1809,163 @@ subroutine init_hist (dt)
+         call define_hist_field(n_siconc,"siconc","1",tstr2D, tcstr,   &
+             "sea area fraction",                                      &
+             "none", c1, c0,                                           &
+             ns1, f_siconc, avg_ice_present=.false., mask_ice_free_points=.true.)
```
What is the criteria for mask_ice_free_points = .true. here? Is this for all variables where avg_ice_present = .true.?

anton-seaice replied:
At the moment we use them both together.

Maybe there is a usecase for native variables to mask ice_free_points without doing the weighted average by aice ?

dabail10 replied:
Certain fields are technically not defined for open ocean points. For example, fsens and flat are zero where there is no ice. However, things like aice and vice have a value of zero where there is no ice and this is a real value. I will try to go through and make these consistent.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1808,15 +1809,163 @@ subroutine init_hist (dt)
+             ns1, f_sisaltmass, avg_ice_present=.false., mask_ice_free_points=.true.)
+
+         call define_hist_field(n_sisnconc,"sisnconc","1",tstr2D, tcstr,   &
+             "sea snow area fraction",                                      &
```
Thanks both - 

I think my point was - what area is the native `snowfrac` calculated over? Presumbaly grid cell area.

So `sisnconc` needs to be `snowfrac/aice` to get sea ice area?

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -655,23 +749,24 @@ subroutine init_hist (dt)
       call broadcast_scalar (f_sidmassgrowthwat, master_task)
       call broadcast_scalar (f_sidmassgrowthbot, master_task)
-      call broadcast_scalar (f_sidmasssi, master_task)
+      call broadcast_scalar (f_sidmassgrowthsi, master_task)
```
We need to warn folks in some way that these variables are deprecated. (Probably warning in code is overkill, but maybe in the docs?)

dabail10 replied:
I have added text in the documentation saying that we have updated these variable names for the CMIP7 data request.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -739,189 +837,194 @@ subroutine init_hist (dt)
-            "ice volume per unit grid cell area", c1, c0,         &
+         call define_hist_field(n_hi,"hi","m",tstr2D, tcstr, &
+            "grid cell mean ice thickness",                  &
+            "ice volume per unit grid cell area", c1, c0,    &
```
Do we have a norm about whitespace only changes? Doing them makes git blame harder to use

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1530,449 @@ subroutine init_hist (dt)
-             "none", c1, c0,         &
-             ns1, f_sicompstren, avg_ice_present=.true., mask_ice_free_points=.true.)
+             "sea-ice thickness",                                      &
+             "volume divided by area", c1, c0,                         &
```
```suggestion
             "volume divided by sea ice area", c1, c0,                         &
```
Something should note this is true thickness

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1530,449 @@ subroutine init_hist (dt)
+             ns1, f_sishearvel)
+
+
+      ! CMIP 2D variables (extensive, avg_ice_present = 'none')
```
I think you are using this to seperate intensive and extensive ? But there are lots of extensive variables above this line

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1530,449 @@ subroutine init_hist (dt)
+             ns1, f_sisnconc)
+
+         call define_hist_field(n_sisnmass,"sisnmass","kg m-2",tstr2D, tcstr, &
+             "snow mass per area",                                            &
```
```suggestion
             "snow mass per grid cell area",                                            &
```

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2501,14 +2667,14 @@ subroutine accum_hist (dt)
 
          if (f_fswabs_ai(1:1)/= 'x') &
-             call accum_hist_field(n_fswabs_ai, iblk, fswabs(:,:,iblk)*workb(:,:), a2D)
+             call accum_hist_field(n_fswabs_ai, iblk, fswabs(:,:,iblk)*aice(:,:,iblk), a2D)
```
I think we should save fswabs_ai seperately (before `scale_fluxes` is called) rather than doing this multiply by `aice`. (There's about 8 more variable below we do the same thing on)

(I guess thats a bit off-topic)

eclare108213 replied:
I agree, but that's for a separate PR.  It's flagged in https://github.com/CICE-Consortium/CICE/issues/67, the oldest issue in the list!

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,21 +2884,140 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) worka(i,j) = vice(i,j,iblk)
+              worka(i,j) = vice(i,j,iblk)
            enddo
            enddo
```
You don't need a loop here any more, just :
```
call accum_hist_field(n_sithick, iblk, vice(:,:,iblk), a2D)
```
is sufficient

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,21 +2884,140 @@ subroutine accum_hist (dt)
+              worka(i,j) = vice(i,j,iblk)
+           enddo
+           enddo
+           call accum_hist_field(n_sivol, iblk, worka(:,:), a2D)
```
```suggestion
           call accum_hist_field(n_sivol, iblk, vice(:,:,iblk), a2D)
```

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,21 +2884,140 @@ subroutine accum_hist (dt)
+              worka(i,j) = aice(i,j,iblk)
+           enddo
+           enddo
+           call accum_hist_field(n_siconc, iblk, worka(:,:), a2D)
```
```suggestion
           call accum_hist_field(n_siconc, iblk, aice(:,:,iblk), a2D)
```

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3007,33 +3293,37 @@ subroutine accum_hist (dt)
+           worka(:,:) = c0
+           do j = jlo, jhi
+           do i = ilo, ihi
+              worka(i,j) = -snoice(i,j,iblk)*rhos/dt
```
```suggestion
              worka(i,j) = -snoice(i,j,iblk)*rhoi/dt
```

Conversion of snow to ice in cice is adiabatic transfer of mass from snow -> ice, so sisndmasssi = -1 * sidmasssi

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3007,33 +3293,37 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = evapi(i,j,iblk)*rhoi
-              endif
+              worka(i,j) = evapi(i,j,iblk)*rhoi
```
Is evapi negative (mostly) ? I think sidmassevapsubl should be less than zero (except for small amounts of deposition/condensation)

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3043,9 +3333,7 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = meltt(i,j,iblk)*rhoi/dt
-              endif
+              worka(i,j) = meltt(i,j,iblk)*rhoi/dt
```
I think this should be multiplied by -1 ?

See from Notz, "Always negative or zero"

> The rate of change of sea-ice mass through melting at the ice surface divided by grid-cell area. This number is independent of the actual fate of the meltwater, and will hence include all sea-ice meltwater that drains into the ocean and all sea-ice meltwater that is collected by a melt-pond parameterisation. Always negative or zero.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3055,9 +3343,7 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = meltb(i,j,iblk)*rhoi/dt
-              endif
+              worka(i,j) = meltb(i,j,iblk)*rhoi/dt
```
*-1 ?

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3067,69 +3353,57 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = aice(i,j,iblk)*fsnow(i,j,iblk)
-              endif
+              worka(i,j) = aice(i,j,iblk)*fsnow(i,j,iblk)
```
Should this be aice_init ?

fsnow is based on coupled snowfall, and therefore from start of timestep ?

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3067,69 +3353,57 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = melts(i,j,iblk)*rhos/dt
-              endif
+              worka(i,j) = melts(i,j,iblk)*rhos/dt
```
*-1 ?

anton-seaice reviewed (commented):
Thanks @dabail10 

I didn't go through everything, but the main themes are:

1. in some places this used fixed values of rhoi, rhos and in others they come from icepack. Is this correct ?
2. Melt processes should always be less than 0 in cmip diagnostics
3. There's lots of loops that aren't needed once `if aice>puny` condition is removed, they can just be `call accum_hist_field(n_..., iblk, ...(:,:,iblk), a2D)`
4. Can all the _ai variables be saved before `scale_fluxes` is called rather than recalculating them

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3248,9 +3505,7 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) then
-                 worka(i,j) = aice(i,j,iblk)*frain(i,j,iblk)
-              endif
+              worka(i,j) = aice(i,j,iblk)*frain(i,j,iblk)
```
Good points. I have removed frain from the siflfwdrain variable.

anton-seaice replied:
> So, for sipr, I guess I should only add this if topo ponds?

I think so? `sipr` should have the fraction of rain entering the melt point (is that all of it?)

dabail10 replied:
This variable is a rate in kg / m^2 s. So, it is the same over the whole grid cell, over the ice, or over the ponds.

dabail10 commented:
> Just added a whole bunch of changes. I think this is getting pretty close. I have a one-month run with the output here. Outstanding issues:
> 
> 1. I have added the CF long_name from the CMIP7 table. Do we want more in the comment field here?
> 2. I'm still not happy with fllwutop. I am getting some excessively large values here at points with very tiny ice concentration (1.0e-8). Maybe this is just what it it is. @eclare108213 and I have been iterating on the math.
> 3. Still working on the siitd fields.

I have fixed the fllwutop field and the siitd fields. Thoughts on what to put in the comments field? The description from the table? the CF standard_name?

dabail10 commented:
> Thanks @dabail10
> 
> I didn't go through everything, but the main themes are:
> 
> 1. in some places this used fixed values of rhoi, rhos and in others they come from icepack. Is this correct ?
> 2. Melt processes should always be less than 0 in cmip diagnostics
> 3. There's lots of loops that aren't needed once `if aice>puny` condition is removed, they can just be `call accum_hist_field(n_..., iblk, ...(:,:,iblk), a2D)`
> 4. Can all the _ai variables be saved before `scale_fluxes` is called rather than recalculating them

- I have fixed all occurrences of rhoi as a constant to use the brine + ice density.
- changed the melt processes to negative
- removed the unnecessary accumulation in worka where appropriate.

dabail10 commented:
> Thanks @dabail10
> 
> I didn't go through everything, but the main themes are:
> 
> 1. in some places this used fixed values of rhoi, rhos and in others they come from icepack. Is this correct ?
> 2. Melt processes should always be less than 0 in cmip diagnostics
> 3. There's lots of loops that aren't needed once `if aice>puny` condition is removed, they can just be `call accum_hist_field(n_..., iblk, ...(:,:,iblk), a2D)`
> 4. Can all the _ai variables be saved before `scale_fluxes` is called rather than recalculating them

On point 4, this is a larger issue. Ultimately we want to get rid of the _ai quantities. However, that is for a future PR.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1530,449 @@ subroutine init_hist (dt)
+
+         call define_hist_field(n_sisnthick,"sisnthick","m",tstr2D, tcstr, &
+             "snow thickness",                                             &
+             "snow volume divided by area", c1, c0,                        &
```
I use the terminology "actual ice thickness" (that's probably what's in the docs)

anton-seaice commented:
> Thoughts on what to put in the comments field? The description from the table? the CF standard_name?

Description field can be quite long ? It's a bit of an annoying thing to hard-code .

I don't have a strong view

eclare108213 commented on cicecore/cicedyn/analysis/ice_history_mechred.F90:
```diff
@@ -182,49 +210,60 @@ subroutine init_hist_mechred_2D
 
+      if (f_sirdgconc(1:1) /= 'x') &
+         call define_hist_field(n_sirdgconc,"sirdgconc","%",tstr2D, tcstr, &
+             "percentage of ridged sea ice",                               &
```
Please clarify whether this is the percent of ice or of grid cell, e.g. "percent of sea ice area that is ridged" (or grid cell area)

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1531,451 @@ subroutine init_hist (dt)
+             ns1, f_sitempbot, avg_ice_present='init', mask_ice_free_points=.true.)
+
+         call define_hist_field(n_siu,"siu","m/s",ustr2D, ucstr, &
+             "x-component of sea-ice velocity",                  &
```
hyphen should not be in x component, but it's okay (optional) in sea-ice here (similarly elsewhere -- see comment about nouns vs adjectives)

dabail10 replied:
In the CMIP table they use a hyphen.

dabail10 commented:
I have used the CMIP7 table long names here. I can certainly add more detail in the description comment to help.

dabail10 commented:
> > sea-ice thickness
> > Actual (floe) thickness of sea ice averaged over the ice-covered part of a given grid cell, NOT volume divided by grid area.
> > ```
> > 2. Should I not have this field since we are doing volume divide by area?
> > ```
> 
> My interpretation is `sivol` is volume divided by grid-cell area, and `sithick` is volume divided by ice area. (Which matches the proposed implementation in this PR)

Right, this description means that models that have sithick as a prognostic variable should write it out. They apparently do not want volume divided by area.

anton-seaice commented:
> They apparently do not want volume divided by area.

Really? Is there a reference for this ?

I thought the description was just trying to clarify its actual thickness rather than "volume per grid-cell area"

dabail10 commented:
I have completely reworked this and removed the f_CMIP flag. I have now added the CMIP variables as a set_nml.cmip option. This gives the users the ultimate flexibility if they wish to depart from the official CMIP data request. Still a bit more to come on this one.

- Add more information in the comments field for the SIMIP variable based on the CMIP data request description field
- Figure out if sithick = sivol / siconc is ok.
- Fix the dashes in sea-ice / sea ice.
- Do I need to add all of the CMIP variable flags to the documentation?

apcraig commented:
> Crap. The instantaneous fields are supposed to be in the middle of the month.

I've been thinking about a lot of features that would be nice to add to history.  One would be an ability to control the "phase" of the output.  So if you want 3 month averages, we cannot control easily whether it's JFM or DJF or maybe Dec 21 - Mar 21.   If you want a daily average, lets say you want it from 12z to 12z the next day.  Or if you want an instantaneous value, maybe you want it on the 15th of the month instead of the 1st.  This and other features would not be trivial to do.  

In the short term as a quick fix, you could probably add a new histfreq option which would allow output to be written on the "nth" of the month.  In ice_calendar, that might look something like

```
         case ("n", "N")
            if (new_day  .and. histfreq_n(ns)/=0) then
               if (mday == histfreq_n(ns)) &
                   write_history(ns) = .true.
            endif
```

This would actually generate a monthly average to that day of the month if hist_avg=true and an instantaneous output if hist_avg=false.  So kind of awkward, but as long as it's documented....

Again, this could be a short term workaround in lieu of adding more robust features which would probably require quite a bit of extra work.

dabail10 commented on cicecore/cicedyn/analysis/ice_history_pond.F90:
```diff
@@ -258,64 +301,89 @@ subroutine init_hist_pond_2D
+      ! CMIP melt pond variables
+      if (f_simpconc(1:1) /= 'x') &
+         call define_hist_field(n_simpconc,"simpconc","%",tstr2D, tcstr, &
+             "fraction of sea-ice covered by melt ponds",                &
```
I think this is fixed now.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
+             "on native model grid", c1, c0,                     &
+             ns1, f_siv, avg_ice_present='final', mask_ice_free_points=.true.)
+
+      ! CMIP 2D variables (intensive, avg_ice_present = 'init' or 'final')
```
```suggestion
```

This comment seems to repeat the above comment

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1890,17 +1909,29 @@ subroutine init_hist (dt)
+              "snow thickness in ice thickness categories",                         &
+              "actual thickness of snow in each ice thickness category",            &
+              c1, c0,                                                               &
+              ns1, f_siitdsnthick, avg_ice_present='final')
```
```suggestion
              ns1, f_siitdsnthick)
```
The divide by aicen should be done every time step, not at once per averaging interval (I dont think they are equivalent?)

Same for siitdthick

dabail10 replied:
I don't understand. These variable siitdthick / siitdsnthick are intensive so I am assuming the time average looks like:

```
sum(vicen(n)) / sum(aicen(n,t)), t = t1,t2 for each n
```

@eclare108213 ?

eclare108213 replied:
it's true, they are not equivalent, but they should have similar values. @dabail10's equation above is the more robust calculation, consistent with the averages documented in https://github.com/CICE-Consortium/CICE/pull/1056. Dividing by aicen at every time step is more intuitive, but mathematically I prefer the more robust approach.

anton-seaice replied:
Sorry - they are intensive, you are correct. :)

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,205 +2754,168 @@ subroutine accum_hist (dt)
-              if (aice(i,j,iblk) > puny) worka(i,j) = aice(i,j,iblk)*trcr(i,j,nt_iage,iblk)
+              if (ktherm == 2) &
+                 call ice_brine_density(trcr(i,j,nt_qice:nt_qice+nzilyr-1,iblk),trcr(i,j,nt_sice:nt_sice+nzilyr-1,iblk), &
+                                        sss(i,j,iblk), rho_ice, rho_ocn, salt_ice)
```
I think it better just to create another array to store rho_ice in, rather than call it multiple times

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,205 +2754,168 @@ subroutine accum_hist (dt)
+         endif
+
+         if (f_sisnconc(1:1) /= 'x') then
+           call accum_hist_field(n_sisnconc, iblk, snowfrac(:,:,iblk), a2D)
```
I still think something is wrong with either sisnconc or snowfrac diagnostic:

`sisnconc` is per sea ice area.
From Notz:
> Area fraction of the sea-ice surface that is covered by snow.

`snowfrac` appears to be per grid cell area
https://github.com/CICE-Consortium/CICE/blob/2760117819011860269a2b3b6a4bb201995cda62/cicecore/cicedyn/analysis/ice_history.F90#L679-L681

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
+             "snow mass change through snowfall",                                           &
+             "rate of change of snow mass due to solid precipitation falling onto sea ice", &
+             c1, c0,                                                                        &
+             ns1, f_sisndmasssnf, avg_ice_present='init')
```
I think set this as mask_ice_free_points=.true. for consistency with the other vars

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3734,73 +3566,66 @@ subroutine accum_hist (dt)
+!                endif
+!             enddo             ! i
+!             enddo             ! j
+!             endif
```
Do we really want a long section of commented out code ?

dabail10 replied:
I guess I am hoping to fix this later.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3734,73 +3566,66 @@ subroutine accum_hist (dt)
                  enddo             ! j
```
When avg_ice_present = 'init' , should we be masking by ravgip_init(i,j) instead of ravgip(i,j)?

We always use avg_ice_present and mask_ice_free_points together. I'd be tempted just to get rid of `mask_ice_free_points` as an option, and then :
- mask by `ravgip_init(i,j) == c0` when `avg_ice_present = 'init'`
- mask by `ravgip(i,j) == c0` when `avg_ice_present = 'final'`

anton-seaice commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_history_write.F90:
```diff
@@ -272,12 +272,13 @@ subroutine ice_write_hist (ns)
                dimid(2) = timid
 
                call ice_hist_coord_def(ncid, time_coord, nf90_double, dimid(1:2), varid)
+               status = nf90_put_att(ncid,varid,'calendar',cal_att) !extra attribute
```
```suggestion
               status = nf90_put_att(ncid,varid,'calendar',cal_att) !extra attribute
               call ice_check_nc(status,  subname//' ERROR: defining att calendar: '//cal_att,file=__FILE__,line=__LINE__)
```

apcraig commented:
> > With these modifications, that test has changed answers but only in the "arwt atm heat flux" diagnostic.
> 
> @apcraig Maybe it's relevant that this branch is a commit behind main [ESCOMP/CICE@cmip7...CICE-Consortium:CICE:main](https://github.com/ESCOMP/CICE/compare/cmip7...CICE-Consortium:CICE:main) and residual ice changed arwt diagnostics too (#645)

My sandbox is Consortium main plus this PR, so it's what should be on the head of main when merged.  And I'm comparing with Consortium main.  But it's a good point that maybe this issue has something to do with #645.  I can't quite think of how, but let me explore that a bit.  Maybe this PR is reverting something that was done in #645.  But why is only the opticep test failing and not all the tests.....  Very odd......

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
+         call define_hist_field(n_sisnhc,"sisnhc","J m-2",tstr2D, tcstr, &
+             "snow heat content",                                        &
+             "heat content of all snow in grid cell", c1, c0,            &
+             ns1, f_sisnhc, avg_ice_present='none', mask_ice_free_points=.true.)
```
I have changed all the extensive variables to remove the masking.

apcraig commented on configuration/scripts/tests/io_suite.ts:
```diff
@@ -3,7 +3,7 @@ smoke          gx3    32x1        run1year,histhrly,ionetcdf,iocdf2,short
 
 # Need to check with cprnc manually, CICE tools always produce diffs between netcdf and pio
-smoke          gx3     8x2        histall,ionetcdf,iocdf5,run5day
+smoke          gx3     8x2        histall,cmip,ionetcdf,iocdf5,run5day
```
I'm not sure this is what we want in the test suite.  histall already turns on all the variables, adding cmip here has little or no impact.  How about if we add a new restart test instead like,

restart        gx3    12x2        cmip,ionetcdf,iocdf2

apcraig commented on configuration/scripts/options/set_nml.histall:
```diff
@@ -237,3 +236,84 @@
     f_aice_ww      = 'md'
     f_diam_ww      = 'md'
     f_hice_ww      = 'md'
+    f_aice             = 'md'
```
f_aice is already set above, please remove this entry.

apcraig replied:
looks like f_icedir and f_icespd needs to be added to set_nml.histall.  Should those be in set_nml.cmip?

dabail10 commented:
> OK, I did a bunch of testing on this difference in the opticep unit test. This PR does modify flwout and that's what we're seeing. The reason the regular tests are bit-for-bit but the opticep unit test is not is that the regular tests are comparing restart files while the opticep is comparing log files. The restart files are bit-for-bit, but the log files are not because the arwt heat (via flwout) has changed. That difference is NOT picked up in our regression testing in general, diagnostics are not part of that comparison except when log files are compared and only the diagnostics written to the log files in that case, not all diagnostics.
> 
> At any rate, I've run a full test suite on this PR and all the tests are performing as expected. I will approve this PR for now. It would be great to get other reviews before we merge.

Sorry, I didn't get to this, but glad you figured it out. I did change how flwout is set in scale_fluxes. I guess I was surprised it only impacted the one test.

apcraig commented:
I will retest the PR once the final changes are made.  Just let me know when things are ready for final testing.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history_mechred.F90:
```diff
@@ -380,6 +407,12 @@ subroutine accum_hist_mechred (iblk)
+                             aice(:,:,iblk)*(c1 - trcr(:,:,nt_alvl,iblk)), a2D)
+         if (f_sirdgthick(1:1)/= 'x') &
+             call accum_hist_field(n_sirdgthick,   iblk, &
+                             vice(:,:,iblk)*(c1 - trcr(:,:,nt_vlvl,iblk)), a2D)
```
This doesn't look like the "total volume of ridged sea ice divided by area of ridges" in line 251 - it's not divided by an area

eclare108213 commented on configuration/scripts/ice_in:
```diff
@@ -613,6 +614,8 @@
     f_icepresent   = 'm'
     f_iage         = 'm'
+    f_icedir       = 'x'
+    f_icespd       = 'x'
```
Why are these here, between the two age flags?  It would make more sense for them to reside next to uvel, vvel or divu.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
+         call define_hist_field(n_siforcecorioly,"siforcecorioly","N m-2",tstr2D, tcstr, &
+             "coriolis force term in force balance (y-component)",                       &
+             "y-component of the force on sea ice caused by the Coriolis force",         &
+             c1, c0,                                                                     &
```
likewise for other terms

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             ns1, f_sidmasslat)
+         call define_hist_field(n_sispeed,"sispeed","m/s",ustr2D, ucstr, &
+             "sea-ice speed",                                            &
+             "speed of ice (i.e. mean absolute velocity)", c1, c0,       &
```
how about "mean velocity magnitude" by itself (without 'speed of ice i.e.')

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             ns1, f_siflswdtop, avg_ice_present=.true., mask_ice_free_points=.true.)
+         call define_hist_field(n_sidmassevapsubl,"sidmassevapsubl","kg m-2 s-1",tstr2D, tcstr, &
+             "sea-ice mass change through evaporation and sublimation",                         &
+             "rate of change of sea-ice mass change through evaporation and sublimation",       &
```
remove the second 'change'

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             "none", c1, c0,                            &
-             ns1, f_sifb, avg_ice_present=.true., mask_ice_free_points=.true.)
+         call define_hist_field(n_simass,"simass","kg m-2",tstr2D, tcstr, &
+             "sea-ice mass per area",                                     &
```
per grid cell area?

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             ns1, f_sifb, avg_ice_present=.true., mask_ice_free_points=.true.)
+         call define_hist_field(n_simass,"simass","kg m-2",tstr2D, tcstr, &
+             "sea-ice mass per area",                                     &
+             "total mass of sea ice", c1, c0,                             &
```
shouldn't this also say "per grid cell area"?

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             ns1, f_siflfwbot, avg_ice_present=.true., mask_ice_free_points=.true.)
+         call define_hist_field(n_sisnmass,"sisnmass","kg m-2",tstr2D, tcstr, &
+             "snow mass per area",                                            &
+             "total mass of snow on sea ice", c1, c0,                         &
```
per grid cell area?

eclare108213 commented:
> Ok. I have modified the computation of the pond and ridge thicknesses. These do look a bit wacky to me. I am getting negative ridge thicknesses. I think this is not correct:
> vice(:,:,iblk) - trcr(:,:,nt_vlvl,iblk)
> I think this should be vice - aice*trcr(:,:,nt_vlvl,iblk)?

No, level-ice volume is a tracer on ice volume, so it should be vice - vice * trcr(nt_vlvl) = vice * (1-trcr(nt_vlvl)).

> For the melt pond thicknesses I am getting some very large values. I think this is because when pond area or ice area gets very small, then the volume conservation gives unrealistic pond depths.

This will continue to be a problem until with fix the issues related to how aice changes during the time step.

dabail10 commented:
> > To figure out what should be included in a layers-of-tracers calculation, I always go back to the integrated, conserved quantity. Looking at just one grid cell at a single point in time, the total pond volume (m^3) will be `sum ( ai * ap * hp * A )` where the sum is over categories, ai is ice area, ap is pond area tracer, hp is the pond thickness (volume per unit area) tracer, and A is the grid cell area. For topo ponds it looks like this, but for level-ice ponds there is an extra term since the pond area tracer is carried on the level-ice area tracer: `sum ( ai * alvl * ap * hp * A )` for level-ice ponds. In the history module, you could use a local variable to specify the pond area as `ap` or `alvl * ap` for topo or level-ice ponds, respectively. Dividing by grid cell area `A` gives you the pond volume per unit grid cell area, aka the mean pond depth over the grid cell, `sum ( ai * ap * hp )` for topo ponds `sum ( ai * alvl * ap * hp )` for level-ice ponds. Dividing by ice area gives you the pond volume per unit ice area, aka the mean pond depth over the ice, `sum ( ai * ap * hp ) / sum ( ai )` for topo ponds `sum ( ai * alvl * ap * hp ) / sum ( ai )` for level-ice ponds. Mean pond depth over the level ice only make sense for level-ice ponds: `sum ( ai * alvl * ap * hp ) / sum ( ai * alvl )`. If you want the mean pond depth over the ponds themselves, then divide by pond area: `sum ( ai * ap * hp ) / sum ( ai * ap )` for topo ponds `sum ( ai * ap * hp ) / sum ( ai * alvl * ap )` for level-ice ponds.
> > Tracking ridged ice is a little more complicated because the tracer is for level ice; ridged ice is diagnostic. Following the same logic as for melt ponds, the mean ridge thickness over the ice area is `sum ( vrdg ) / sum ( ai ) = sum ( vi - vlvl ) / sum ( ai )` where vrdg is the ridged-ice volume, vi is the total ice volume and vlvl is the level-ice volume (all per unit grid cell area). The mean ridge thickness over the ridged-ice area would be `sum ( vrdg ) / sum ( ardg ) = sum ( vi - vlvl ) / sum ( ai * (1-alvl) )`
> > I hope that all makes sense.
> 
> I believe the latest changes do this correctly and I have also changed the pond avg_ice_present to use aice_init.

The issue I am coming up against here is that when there is very small ice area and very small ridged / pond area. You are dividing by very small numbers.

simpthick = sum(ai*alvl*ap*hp) / sum(ai*alvl*ap)

sirdgthick = sum(vrdg(t)) / sum(ardg(t))

This is similar to sithick

sithick = sum(vice(t)) / sum(aice(t))

These are kind of artificial properties in CICE. Anyhow, I sort of "fixed" sirdgthick and get maximum values of 40m. This still seems unrealistic, but then we do get sithick values of 26m.

eclare108213 commented:
@dabail10 Is this PR ready for final review?

eclare108213 commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -1396,17 +1396,20 @@ subroutine **define_hist_field**.  ``cona`` and ``conb`` are multiplicative and
+
+The ``f_CMIP`` flag has been removed. This is now a ``set_nml.cmip`` namelist option
+which can be invoked with the ``-s cmip`` option during cice.setup. This optional
+namelist setting will turn on the CMIP data request and turn of CICE duplicates of
```
change 'of' to 'off'

eclare108213 commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -1396,17 +1396,20 @@ subroutine **define_hist_field**.  ``cona`` and ``conb`` are multiplicative and
+namelist setting will turn on the CMIP data request and turn of CICE duplicates of
+SIMIP variables. However, these can be changed by the user in their case ``ice_in`` file.
+Note that all SIMIP variables have been updated to correspond to the new 
+`CMIP7 data request <https://wcrp-cmip.org/cmip7-data-request-v1-0/>`_. 
```
This might be the best place to add a sentence like 

Some SIMIP variables require division by ice or sub-ice areas, which can be extremely small and cause the output variables themselves to appear unphysically large.  Such edge cases can be masked using ...

(if you decide to create a namelist option for history area masking).

eclare108213 commented on configuration/scripts/options/set_nml.cmip:
```diff
@@ -0,0 +1,143 @@
+    f_sisnmass         = 'm'
+    f_sispeed          = 'm'
+    f_sistressave      = 'n'
+    f_sistressmax      = 'n'
```
So 'n' means instantaneous?  Is that documented?  We used to just have those fields hardwired because averages are usually not computed correctly and can be very misleading.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history_pond.F90:
```diff
@@ -456,58 +507,36 @@ subroutine accum_hist_pond (iblk)
-                                                  * trcr(:,:,nt_ipnd,iblk), a2D)
-         endif ! ponds
+         if (f_simpconc(1:1)/= 'x') &
+             call accum_hist_field(n_simpconc, iblk, worka(:,:), a2D)
```
Is simpconc divided by sum(ice area) later, along with all the other variables?  If so, then ice area should be multiplied here.  That would make it consistent with all the other variables too.

eclare108213 replied:
So `avg_ice_present = 'none'`?

eclare108213 commented on cicecore/drivers/nuopc/cmeps/ice_import_export.F90:
```diff
@@ -1073,6 +1073,7 @@ subroutine ice_export( exportState, rc )
        call state_setexport(exportState, 'Si_imask', input=ocn_gridcell_frac, rc=rc)
        if (ChkErr(rc,__LINE__,u_FILE_u)) return
     else
+       tempfld(:,:,:) = c0
```
likewise for snow

eclare108213 replied:
I don't see anything about this in the header.  It looks like a bug fix to me.  Is it?

dabail10 replied:
I don't think it is technically a bug fix as this is done for all internal cells. I just thought is was being careful to initialize everything to zero including the ghost cells even though these are not passed to the coupler.

eclare108213 commented on cicecore/cicedyn/infrastructure/io/io_pio2/ice_history_write.F90:
```diff
@@ -404,15 +405,14 @@ subroutine ice_write_hist (ns)
+      var_nverts(n_latu_bnds) = coord_attributes('latu_bounds','latitude bounds (U-cell)','degrees_north','und')
+      var_nverts(n_lonn_bnds) = coord_attributes('lonn_bounds','longitude bounds (N-cell)','degrees_east','und')
+      var_nverts(n_latn_bnds) = coord_attributes('latn_bounds','latitude bounds (N-cell)','degrees_north','und')
+      var_nverts(n_lone_bnds) = coord_attributes('lone_bounds','longitude bounds (E-cell)','degrees_east','und')
```
Do we use the 'N-cell' and 'E-cell' terminology anywhere else?  Is that specified by CF rules?

dabail10 replied:
Here are the comments from ice_grid.F90:

        lonn_bounds, & ! longitude of gridbox corners for N point
         latn_bounds, & ! latitude of gridbox corners for N point
         lone_bounds, & ! longitude of gridbox corners for E point
         late_bounds    ! latitude of gridbox corners for E point

Maybe I should say corners here.

apcraig replied:
I think setting bounds attributes can cause problems with CF compliance.  I'm not entirely sure what the requirement is.  At a minimum, the attributes must match the variable they are associated with exactly.  If the attributes are not defined, they will be inherited from the associated variable.  I think I ran into some issues with the CF checkers with the bounds attributes.  We should definitely run the updated history files through the CF checkers before merging just so we know where we stand.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history_mechred.F90:
```diff
@@ -182,49 +195,63 @@ subroutine init_hist_mechred_2D
 
+      if (f_sirdgconc(1:1) /= 'x') &
+         call define_hist_field(n_sirdgconc,"sirdgconc","%",tstr2D, tcstr, &
+             "percentage of ridged sea ice",                               &
```
Instead of "percentage of ridged sea ice", which is confusing at first glance because it sounds like you're getting the percentage of something else (not sure what) in ridged ice, how about "ridged sea ice percentage"?  If this language is specified by CMIP7, then leave it...

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2330,8 +2361,8 @@ subroutine accum_hist (dt)
-      !$OMP             k,n,qn,ns,sn,rho_ocn,rho_ice,Tice,Sbr,phi,rhob,dfresh,dfsalt,sicen, &
-      !$OMP             worka,workb,worka3,Tinz4d,Sinz4d,Tsnz4d)
+      !$OMP             k,n,qn,ns,sn,rho_ocn,rho_ice,sal_ice,dfresh,dfsalt,sicen, &
+      !$OMP             worka,workb,Tinz4d,Sinz4d,Tsnz4d)
```
Are the OMP changes intentional?  Are they fixing something?

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2448,7 +2481,7 @@ subroutine accum_hist (dt)
          if (f_fswup(1:1) /= 'x') &
             call accum_hist_field(n_fswup, iblk, &
-                 (fsw(:,:,iblk)-fswabs(:,:,iblk)*workb(:,:)), a2D)
+                 (fsw(:,:,iblk)-fswabs(:,:,iblk)*aice(:,:,iblk)), a2D)
```
In the original code, workb = aice_init.  Just making sure that this change to aice is intentional!

eclare108213 replied:
Then the same bug fix also applied to flat, fsens, flowout, evap, fsurf and fcondtop?  That’s major and should also be noted in the PR header for incorporation into the merge commit.

dabail10 replied:
Duly noted.

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,121 +2753,105 @@ subroutine accum_hist (dt)
-              endif
-           enddo
-           enddo
-           call accum_hist_field(n_sidir, iblk, worka(:,:), a2D)
```
Where is sidir calculated now?

eclare108213 commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -1402,15 +1402,21 @@ Project (SIMIP) :cite:`Notz16` are available as history output variables
 Note that all SIMIP variables have been updated to correspond to the new 
 `CMIP7 data request <https://wcrp-cmip.org/cmip7-data-request-v1-0/>`_. 
 
+Note that some SIMIP variables require division ice or sub=ice areas, which can be extremely
```
2 typos:  "division by ice or sub-ice areas"

Also, please enclose "such as ``sithick``" either in commas or parentheses.

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1427,396 +1359,483 @@ subroutine init_hist (dt)
-             ns1, f_sidmasstrany)
+      ! CMIP 2D variables (for "intensive" variables per Notz et al 2016 definition, 
+      ! use avg_ice_present = 'init' or 'final' to divide by sum(aice) over time, when 
+      ! aice is at the start of the timestep ('init') or the end of the timestep ('final')
```
Ok. I have explicitly added avg_ice_present='none', mask_ice_free_points=.false. to all the extensive CMIP calls.

dabail10 commented:
Just found the Conversations tab! Holy moly. I have resolved all of @anton-seaice 's comments. There is one remaining conversation about the latitude / longitude bounds. @apcraig @eclare108213

apcraig commented:
> Just found the Conversations tab! Holy moly. I have resolved all of @anton-seaice 's comments. There is one remaining conversation about the latitude / longitude bounds. @apcraig @eclare108213

I'm fine with the lat/lon bounds attributes.  I just want to make sure we pass the history files through a CF checker or two.  We don't have to fix everything but it'd be good to know how close we are to meeting the standard and be clear if there are cases where we don't want to meet the CF convention.

dabail10 commented:
> > Just found the Conversations tab! Holy moly. I have resolved all of @anton-seaice 's comments. There is one remaining conversation about the latitude / longitude bounds. @apcraig @eclare108213
> 
> I'm fine with the lat/lon bounds attributes. I just want to make sure we pass the history files through a CF checker or two. We don't have to fix everything but it'd be good to know how close we are to meeting the standard and be clear if there are cases where we don't want to meet the CF convention.

I did run it through a CF checker. Here is the warnings:

WARN: (7.1): Boundary var time_bounds should not have attribute units
WARN: (7.1): Boundary var time_bounds should not have attribute calendar

eclare108213 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -1455,9 +1456,11 @@ subroutine init_hist (dt)
+      ! Use avg_ice_present = 'init', 'final', 'pond', or 'ridge' to divide by 
       ! sum(aice), sum(apond), or sum(ardg) over time 
       ! aice is at the start of the timestep ('init') or the end of the timestep ('final')
+      ! avg_ice_present = 'none' means a standard time average
```
Sorry, maybe too many holiday cookies are fuzzing up my brain, but I don't understand this description.  

In "accumulate grid cell mean values as they are divided by...", does "as" mean "because" or "simultaneously"?  Maybe take that line out altogether (as it was). 

Second, what is a "standard" time average?  The divisions by ice area are for spatial averaging, not time averaging. I suggest
> ! average_ice_present = 'none' does not divide by ice area

dabail10 replied:
The division by ice area is a time average when there is ice present.

dabail10 replied:
How does this sound?

      ! CMIP 2D variables (for "intensive" variables per Notz et al 2016 definition,
      ! that is a weighted time average when ice is present.
      ! Use avg_ice_present = 'init', 'final', 'pond', or 'ridge' to divide by
      ! sum(aice), sum(apond), or sum(ardg) over time
      ! aice is at the start of the timestep ('init') or the end of the timestep ('final')
      ! avg_ice_present = 'none' means a simple time average

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3018,8 +3018,8 @@ subroutine accum_hist (dt)
-              if (fsw(i,j,iblk) > puny .and. aice_init(i,j,iblk) > puny) then
-                 worka(i,j) = aice(i,j,iblk)*(fsw(i,j,iblk)-fswabs(i,j,iblk)*aice(i,j,iblk)/aice_init(i,j,iblk))
+              if (fsw(i,j,iblk) > puny) then
+                 worka(i,j) = (fsw(i,j,iblk)*aice_init(i,j,iblk)-fswabs(i,j,iblk)*aice(i,j,iblk))
```
When aice = 0, then fswabs is zero and if aice_init > 0, then you will get basically that fswup = aice_init*fsw, or basically all of the incoming shortwave over the ice is reflected back. If aice >0 but aice_init = 0, then fswabs = 0 (because there was no ice at the beginning of the step) and fswup = 0. I could add another if condition to make the first case also zero.

anton-seaice reviewed (approved):
Looks good @dabail10 - thanks for all your work. Left a couple if fairly unimportant stylistic comments

apcraig commented:
I have run a full test suite on derecho with 6 compilers and results are bit-for-bit and as expected otherwise.  Right now, the PR looks good.  I have not reviewed the history file results or new diagnostics in any detail.   I will run an appropriate subset of the suite again if there are further changes before merging.

anton-seaice commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -2718,121 +2768,122 @@ subroutine accum_hist (dt)
-           enddo
-           enddo
-           call accum_hist_field(n_sisnthick, iblk, worka(:,:), a2D)
+           call accum_hist_field(n_sisnthick, iblk, vsno(:,:,iblk), a2D)
```
From notz:

> Thickness of snow averaged over the snow-covered part of the sea ice. It hence describes the actual thickness of the snow, which in the context of the heat budget is for example necessary to analyse the heat flux through the ice and snow.

Shouldn't this be `vsno/snowfrac` to give true thickness? (Did we already have this discussion !?)

anton-seaice replied:
Yeah - so doesn't that give the average snow thickness over all sea ice in the cell, but we want the snow thickness only where the is snow on sea ice? e.g. 

 it gives _(snow volume)/(sea ice area)_ and the request is for  _(snow volume)/(area of snow on sea ice)_

eclare108213 replied:
> Yeah, I wish there was a `sisnvol` variable so this could be clear.

Maybe this is a solution to the conundrum?  Even if this variable isn't in the CMIP request, you could add it for clarity.  Otherwise, I suggest keeping @dabail10 's approach, even though it's inconsistent with the pond output.

dabail10 replied:
I disagree about sisnmass. From the data request:

Total mass of snow on sea ice divided by grid-cell area. This then allows one to analyse the storage of latent heat in the snow, and to calculate the snow-water equivalent.

This says nothing about snow fraction. So, it is rhos * vsno. Anyhow. Is this fine how it is for sisnthick?

anton-seaice replied:
I don't have an answer here

The Notz description explicitly says:

_It can also be derived by dividing the total volume of snow through the area of the snow._

But that has no prognostic meaning (i.e. conductive heat flux is based on the total volume of snow divided by area of sea-ice ?).

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3712,6 +3452,39 @@ subroutine accum_hist (dt)
+                                      * ravgip_ridge(i,j) + avail_hist_fields(n)%conb
+                    endif
+                 enddo             ! i
+                 enddo             ! j
```
Good idea for the refactoring.

dabail10 commented on cicecore/cicedyn/analysis/ice_history.F90:
```diff
@@ -3683,6 +3390,39 @@ subroutine accum_hist (dt)
            enddo             ! i
            enddo             ! j
            endif
+           if (n_aice_init(ns) > 0) then
```
This is probably a good idea.

dabail10 commented:
I realize my abort conditions were not flexibile enough. I have added conditionals like:

```
 if (n_ardg(ns) > 0 .and. any(avail_hist_fields(:)%avg_ice_present == 'ridge')) then
```

anton-seaice commented:
I still think there is something inconsistent with snowfrac / sisnconc:

Snowfrac is per grid cell area

https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L686-L689

Sisnconc is per sea ice area (and extensive)

https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L1780-L1784

But they are calculated the same:

https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L2401-L2402

https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L2818-L2820

dabail10 commented:
> I still think there is something inconsistent with snowfrac / sisnconc:
> 
> Snowfrac is per grid cell area
> 
> https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L686-L689
> 
> Sisnconc is per sea ice area (and extensive)
> 
> https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L1780-L1784
> 
> But they are calculated the same:
> 
> https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L2401-L2402
> 
> https://github.com/CICE-Consortium/CICE/blob/d0455f644a068b8daa9f4ec69c829691c7f1791d/cicecore/cicedyn/analysis/ice_history.F90#L2818-L2820

Ah, I think I understand. The variable snowfrac is only the fraction on the sea ice. This comment is incorrect.

dabail10 commented:
Getting a lot of compare fails.

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#b653260a0f28509a80598e07635d01d0c56ce434

apcraig commented:
So, are there five issues that should close?

"To automatically close issues when merging a pull request (PR), use specific keywords like
closes, fixes, or resolves followed by the issue number (e.g., closes #123) in the PR's description or commit message, which triggers the closure upon merging into the default branch on platforms like GitHub and GitLab."

It looks like you added the issues that are related, please update the PR description to explicitly add phrasing like

closes #123 

for any issue that should automatically close.

apcraig commented:
> Added. I thought if they were linked it would do the same.

Maybe they do, it's hard to know for sure.  Thanks for adding that text.
